packages/repository/src/relations/relation.types.ts

```diff
@@ -87,12 +87,12 @@ export interface HasManyThroughDefinition extends RelationDefinitionBase {
   /**
    * The foreign key in the source model, e.g. Customer#id.
    */
-  keyFrom: string;
+  keyFrom?: string;
 
   /**
    * The primary key of the target model, e.g Seller#id.
    */
-  keyTo: string;
+  keyTo?: string;
```

Comment:
While we are discussing `HasManyThroughDefinition`. I am not sure if it's actually a good idea to have such interface, see the discussion in https://github.com/strongloop/loopback-next/pull/2359#discussion_r317011708 for more context.

Having two different interfaces for hasMany and hasManyThrough relations is already creating extra complexity as can be seen in my comment above related to `@hasMany()` decorator.

Cross-posting from https://github.com/strongloop/loopback-next/pull/2359#discussion_r317011708:

> So far, we are using `type` to discriminate between different interfaces inheriting from `RelationDefinitionBase`. See https://www.typescriptlang.org/docs/handbook/advanced-types.html#discriminated-unions
>
> If we want to keep using `hasMany` value for has-many-through relations, then I am fine with that, but let's capture that in the type system too.
>
> We can either get rid of `HasManyThroughDefinition` completely and move all members to `HasManyDefinition`, or else modify `HasManyThroughDefinition` to inherit from `HasManyDefinition`.

Personally, I like most the option where there is a single `HasManyDefinition` interface with an optional `through` field that's used to distinguish between hasMany and hasManyThrough cases. Similarly, `HasManyResolvedDefinition` can contain an optional `through` field that's used to distinguish hasManyThrough case, this field can use a type like `ThroughResolvedDefinition` to tell the compiler that optional properties were resolved and will be always present.

I think it would be best to extract the changes in relation definitions, the implementation bits of resolving through relations and the accompanying unit tests into a new pull request, to get a smaller patch that's easier to review & iterate on.